cs_volume.py

```diff
@@ -0,0 +1,749 @@
+            args['domainid'] = self.get_domain(key='id')
+            args['projectid'] = self.get_project(key='id')
+            args['listall'] = True
+
+            volumes = self.cs.listVolumes(**args)
+
+            if volumes:
+                volume_name = self.module.params.get('name')
+                for v in volumes['volume']:
+                    if volume_name.lower() == v['name'].lower():
+                        self.volume = v
+                        break
+        return self.volume
+
+    def present_volume(self):
+        volume = self.get_volume()
+
+        if not volume:
+            self.result['changed'] = True
+
+            args = {}
+            args['name'] = self.module.params.get('name')
+            args['account'] = self.get_account(key='name')
+            args['domainid'] = self.get_domain(key='id')
+            args['diskofferingid'] = self.get_disk_offering(key='id')
+            args['displayvolume'] = self.module.params.get('display_volume')
+            args['maxiops'] = self.module.params.get('max_iops')
+            args['miniops'] = self.module.params.get('min_iops')
+            args['projectid'] = self.get_project(key='id')
+            args['size'] = self.module.params.get('size')
+            args['zoneid'] = self.get_zone(key='id')
+
+            if not self.module.check_mode:
+                res = self.cs.createVolume(**args)
+                if 'errortext' in res:
+                    self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+                poll_async = self.module.params.get('poll_async')
+                if poll_async:
+                    volume = self.poll_job(res, 'volume')
+        return volume
+
+    def attached_volume(self):
+        missing_params = []
+
+        if missing_params:
+            self.module.fail_json(msg="missing required arguments: %s" % ','.join(missing_params))
+
+        volume = self.present_volume()
+
+        if 'virtualmachineid' in volume and volume['virtualmachineid'] != self.get_vm(key='id'):
+            self.result['changed'] = True
+
+            if not self.module.check_mode:
+                volume = self.detached_volume()
+
+        if 'attached' not in volume:
+            self.result['changed'] = True
+
+            args = {}
+            args['id'] = volume['id']
+            args['virtualmachineid'] = self.get_vm(key='id')
+            args['deviceid'] = self.module.params.get('device_id')
+
+            if not self.module.check_mode:
+                res = self.cs.attachVolume(**args)
+                if 'errortext' in res:
+                    self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+                poll_async = self.module.params.get('poll_async')
+                if poll_async:
+                    volume = self.poll_job(res, 'volume')
+        return volume
+
+    def detached_volume(self):
+        volume = self.get_volume()
+
+        if volume:
```

Comment:
Mustn't we check if the volume is attached here too?
~~~
if volume and and 'attached' in volume:
~~~